web/codechecker_web/shared/util.py

```diff
@@ -0,0 +1,390 @@
+        return ex.output, ex.returncode
+    except OSError as oerr:
+        LOG.warning(oerr.strerror)
+        return oerr.strerror, oerr.errno
+
+
+def get_default_workspace():
+    """
+    Default workspace in the users home directory.
+    """
+    workspace = os.path.join(os.path.expanduser("~"), '.codechecker')
+    return workspace
+
+
+def __strip_protocol_from_url(url):
+    """
+    Strip the PROTOCOL specifier from an URL string and return it, along with
+    the remainder of the URL.
+    """
+    if '://' not in url:
+        return None, url
+
+    parts = url.split('://', 1)
+    protocol = parts[0]
+    url = url.replace(parts[0] + '://', '').lstrip('/').rstrip('/')
+
+    return protocol, url
+
+
+def expand_whole_protocol_and_port(protocol=None, port=None):
+    """
+    Calculate a full protocol and port value from the possibly None protocol
+    and port values given. This method helps to default port numbers to
+    connection protocols understood by CodeChecker.
+    """
+
+    if protocol:
+        if protocol == 'http':
+            portnum = 80
+        elif protocol == 'https':
+            portnum = 443
+        else:
+            raise ValueError("'{0}' is not a protocol understood by "
+                             "CodeChecker".format(protocol))
+    else:
+        protocol = 'http'
+        portnum = 8001
+
+    if port:
+        portnum = port
+
+    return protocol, portnum
+
+
+def split_server_url(url):
+    """
+    Splits the given CodeChecker server URL into its parts.
+
+    The format of a valid URL is:
+      protocol://host:port/
+    where
+      * Protocol: HTTP or HTTPS
+      * Host: The server's host name or IP address
+      * Port: The server's port number
+
+    As a shortcut, the following formats are also valid:
+      hostname           (means: http://hostname:8001)
+    """
+
+    LOG.debug("Parsing server url '%s'", url)
+    protocol, url = __strip_protocol_from_url(url)
+
+    # A valid product_url looks like this: 'http://localhost:8001/Product'.
+    protocol, port = expand_whole_protocol_and_port(protocol, None)
+    host = 'localhost'
+    try:
+        parts = url.split('/', 1)
+
+        # Something is either a hostname, or a host:port.
+        server_addr = parts[0].split(":")
+        if len(server_addr) == 2:
+            host, port = server_addr[0], int(server_addr[1])
+        elif len(server_addr) == 1:
+            host = server_addr[0]
+        else:
+            raise ValueError("The server's address is not in a valid "
+                             "'host:port' format!")
+    except Exception:
+        raise ValueError("The specified server URL is invalid.")
+
+    LOG.debug("Result: With '%s' on server '%s:%s'",
+              protocol, host, port)
+
+    return protocol, host, port
+
+
+def create_product_url(protocol, host, port, endpoint):
+    return "{0}://{1}:{2}{3}".format(protocol, host, str(port), endpoint)
+
+
+def split_product_url(url):
```

Comment:
`web/client/codechecker_client/product.py`? With all the corresponding product url handling, which is used by the client.